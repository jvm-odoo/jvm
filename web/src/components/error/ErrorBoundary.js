import React, { Component } from 'react'

import errorImage from '../../assets/img/error.svg'
import {
  Container,
  Section,
  Title,
  Text,
} from '../common'
import { ErrorImage, ErrorDetailsSummary } from './'

export class ErrorBoundary extends Component {
  constructor(props) {
    super(props)
    this.state = {
      error: null,
      errorInfo: null,
      isDebug: false,
      isMobile: window.innerWidth < 768,
    }
  }

  componentDidCatch(error, errorInfo) {
    this.setState({
      error: error,
      errorInfo: errorInfo,
      isDebug: window.location.search.indexOf('debug') !== -1,
    })
  }

  refresh(e) {
    e.preventDefault()
    localStorage.clear()
    window.location.reload()
  }

  render() {
    if (this.state.errorInfo) {
      return (
        <Container>
          <Section
            column
            horizontallyCentered
            verticallyCentered
            mobilePadding
          >
              <ErrorImage src={errorImage} alt="Error" />
              <Title>Oops!</Title>
              <Text margin centered>
                Something went wrong. {this.state.isMobile && <br />}Try to <a href='/' onClick={this.refresh}>click here</a>.<br/>
                If the problem persists, {this.state.isMobile && <br />}<a href='mailto:jasonvanmalder@gmail.com'>please contact me</a>.
              </Text>
              {this.state.isDebug && (
                <details style={{ whiteSpace: 'pre-wrap' }}>
                  <ErrorDetailsSummary><Text>Details</Text></ErrorDetailsSummary>
                  <pre>
                    {this.state.error && this.state.error.toString()}
                    <br />
                    {this.state.errorInfo.componentStack}
                  </pre>
                </details>
              )}
          </Section>
        </Container>
      )
    }
    return this.props.children
  }
}
