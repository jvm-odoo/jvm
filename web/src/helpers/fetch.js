export const fetchGet = async (url) => {
  const response = await fetch(url)
  return response.json()
}
