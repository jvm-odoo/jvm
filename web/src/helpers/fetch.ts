export const fetchGet = async (url: string) => {
    const response = await fetch(url)
    return response.json()
}

export const fetchPost = async (url: string, data: any, token?: string) => {
    const headers: Headers = new Headers({
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*'
    })

    if (token) {
        headers.append('Authorization', `Token ${token}`)
    }

    const response = await fetch(url, {
        method: 'POST',
        headers,
        mode: 'cors',
        body: JSON.stringify(data)
    })

    return response.json()
}