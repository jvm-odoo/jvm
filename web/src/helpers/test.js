import React from 'react'
import toJson from 'enzyme-to-json'
import { act } from 'react-dom/test-utils'
import { mount, shallow } from 'enzyme'
import { MockedProvider } from '@apollo/react-testing'
import { Provider } from 'react-redux'
import { ThemeProvider } from 'styled-components'

import { store } from '../redux/store'
import { THEME_LIGHT } from '../constants'

const wait = (time = 0) => new Promise(res => setTimeout(res, time))

export const executeMockProviderTestCase = async (wrapperInstance) => (
  await act(async () => {
    await wait(100)
    wrapperInstance.update()
  })
)

export const shallowMatchSnapshot = (Component) => {
  const wrapper = shallow(
    <ThemeProvider theme={THEME_LIGHT}>
      {Component}
    </ThemeProvider>)
  expect(toJson(wrapper)).toMatchSnapshot()
}

export const mountWrapperWithMockedData = async (Component, mocks) => {
  let wrapper

  act(() => {
    wrapper = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ThemeProvider theme={THEME_LIGHT}>
          {Component}
        </ThemeProvider>
      </MockedProvider>
    )
  })

  await executeMockProviderTestCase(wrapper)
  return wrapper
}

export const withReduxProvider = (Component) => (
  <Provider store={store}>
    {Component}
  </Provider>
)
