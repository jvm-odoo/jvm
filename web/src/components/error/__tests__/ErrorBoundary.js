import React from 'react'
import toJson from 'enzyme-to-json'
import { act } from 'react-dom/test-utils'
import { mount } from 'enzyme'

import { ErrorBoundary } from '../ErrorBoundary'
import { executeMockProviderTestCase } from '../../../helpers/test'

it('renders correctly an error', async () => {
  let wrapper

  act(() => {
    wrapper = mount(<ErrorBoundary><div id='prout'/></ErrorBoundary>)
  })

  wrapper.setState({
    error: 'Very big error',
    errorInfo: { componentStack: 'prout' },
    isDebug: true,
  })

  await executeMockProviderTestCase(wrapper)
  expect(wrapper.text().indexOf('Something went wrong')).not.toBe(-1)
  expect(wrapper.text().indexOf('Details')).not.toBe(-1)
  expect(toJson(wrapper)).toMatchSnapshot()
})

