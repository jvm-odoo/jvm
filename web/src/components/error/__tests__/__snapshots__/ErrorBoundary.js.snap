// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly an error 1`] = `
<ErrorBoundary>
  <styled.div>
    <div
      className="sc-AxjAm cJWZhD"
    >
      <styled.section
        column={true}
        horizontallyCentered={true}
        mobilePadding={true}
        verticallyCentered={true}
      >
        <section
          className="sc-AxgMl epjfXB"
        >
          <styled.img
            alt="Error"
            src="error.svg"
          >
            <img
              alt="Error"
              className="sc-AxmLO bmcvVM"
              src="error.svg"
            />
          </styled.img>
          <styled.h1>
            <h1
              className="sc-AxiKw drZJMY"
            >
              Oops!
            </h1>
          </styled.h1>
          <styled.p
            centered={true}
            margin={true}
          >
            <p
              className="sc-AxhCb fVUwQN"
            >
              Something went wrong. 
              Try to 
              <a
                href="/"
                onClick={[Function]}
              >
                refresh the page
              </a>
              .
              <br />
              If the problem persists, 
              <a
                href="mailto:jasonvanmalder@gmail.com"
              >
                please contact me
              </a>
              .
            </p>
          </styled.p>
          <details
            style={
              Object {
                "whiteSpace": "pre-wrap",
              }
            }
          >
            <styled.summary>
              <summary
                className="sc-fzozJi dMxRXQ"
              >
                <styled.p>
                  <p
                    className="sc-AxhCb borSJa"
                  >
                    Details
                  </p>
                </styled.p>
              </summary>
            </styled.summary>
            <pre>
              Very big error
              <br />
              prout
            </pre>
          </details>
        </section>
      </styled.section>
    </div>
  </styled.div>
</ErrorBoundary>
`;
